test/unit/Controller/ApiControllerTest.php

```diff
@@ -0,0 +1,331 @@
+
+        $apiController = new ApiController(
+            $this->specificationLoader,
+            $this->router,
+            $this->serializer,
+            $this->eventDispatcher,
+            $this->requestValidator
+        );
+        $apiController->setApiLoader($apiLoader);
+
+        $this->expectException(Throwable::class);
+        $this->expectExceptionMessage('Api call implementation not found. Please implement "test_function_string" interface');
+
+        $request = new Request(
+            [],
+            [],
+            ['_route_params' => '', '_route' => 'test'],
+        );
+        $apiController->handle($request);
+    }
+
+    public function testHandleWithSetRequestRequestHandler(): void
+    {
+        $apiLoader = new class implements ApiLoader {
+            public function get(string $interfaceName): ?RequestHandler
+            {
+                return new class implements SetRequest, RequestHandler {
+                    public function setRequest(Request $request): void
+                    {
+                        throw new Exception('test request was set');
+                    }
+                };
+            }
+
+            /**
+             * @return string[]
+             */
+            public static function getSubscribedServices(): array
+            {
+                return ['operation_name' => 'test_function_string'];
+            }
+        };
+
+        $routeCollection = new RouteCollection();
+        $routeCollection->add('test', new Route('/test', [], [], ['_openapi_operation' => 'test_option']));
+        $this->router->method('getRouteCollection')->willReturn($routeCollection);
+        $this->specificationLoader->method('load')->willReturn(new Specification(['test_option' => new Operation('/', 'test', 'operation_name')], new OpenApi([])));
+
+        $apiController = new ApiController(
+            $this->specificationLoader,
+            $this->router,
+            $this->serializer,
+            $this->eventDispatcher,
+            $this->requestValidator
+        );
+        $apiController->setApiLoader($apiLoader);
+
+        $this->expectException(Throwable::class);
+        $this->expectExceptionMessage('test request was set');
+
+        $request = new Request(
+            [],
+            [],
+            ['_route_params' => '', '_route' => 'test'],
+        );
+        $apiController->handle($request);
+    }
+
+    public function testHandleWithSetClientIpRequestHandler(): void
+    {
+        $apiLoader = new class implements ApiLoader {
+            public function get(string $interfaceName): ?RequestHandler
+            {
+                return new class implements SetClientIp, RequestHandler {
+                    public function setClientIp(string $ip): void
+                    {
+                        throw new Exception('test ip was set');
```

Comment:
This can be tested without exceptions, by creating a getter on this class and asserting that it returns the expected ip address after the controller is executed.